vignettes/elpd.Rmd

```diff
@@ -0,0 +1,213 @@
+
+## Splitting the data between train and test 
+
+The data is divided between train (80% of the data) and test (20%):
+
+```{r modelfit-holdout, message=FALSE}
+# Prepare data
+data(roaches, package = "rstanarm")
+roaches$roach1 <- sqrt(roaches$roach1)
+roaches$offset <- log(roaches[,"exposure2"])
+# 20% of the data goes to the test set:
+roaches$test <- 0
+roaches$test[sample(.2 * seq_len(nrow(roaches)))] <- 1
+# data to "train" the model
+data_train <- list(y = roaches$y[roaches$test == 0],
+                   x = as.matrix(roaches[roaches$test == 0,
+                                         c("roach1", "treatment", "senior")]),
+                   N = nrow(roaches[roaches$test == 0,]),
+                   K = 3,
+                   offset = roaches$offset[roaches$test == 0],
+                   beta_prior_scale = 2.5,
+                   alpha_prior_scale = 5.0
+                   )
+# data to "test" the model
+data_test <- list(y = roaches$y[roaches$test == 1],
+                   x = as.matrix(roaches[roaches$test == 1,
+                                         c("roach1", "treatment", "senior")]),
+                   N = nrow(roaches[roaches$test == 1,]),
+                   K = 3,
+                   offset = roaches$offset[roaches$test == 1],
+                   beta_prior_scale = 2.5,
+                   alpha_prior_scale = 5.0
+                   )
+```
+
+## Fitting the model to the with RStan
+
+Next we fit the model to the "test" data in Stan using the __rstan__ package:
+
+```{r fit-train}
+# Compile
+stanmodel <- stan_model(model_code = stancode)
+# Fit model
+fit <- sampling(stanmodel, data = data_train, seed = seed, refresh = 0)
+```
+And we draw samples from the generated quantities block based on the "train" data:
+
+```{r gen-test}
+gen_test <- gqs(stanmodel, draws = as.matrix(fit), data= data_test)
+log_lik_test <- extract_log_lik(gen_test)
+```
+
+## Computing holdout elpd:
+
+Now we evaluate the predictive performance of the model on the test data using `elpd()`.
+
+```{r elpd-holdout}
+(elpd_holdout <- elpd(log_lik_test))
+```
+
+When one wants to compare different models, the function `loo_compare()` can be used to assess the difference in performance.
+
+# K-fold cross validation
+
+For this approach the data is divided into folds, and each time one fold is tested while the rest of the data is used to fit the model (see Vehtari et al., 2017).
+
+## Splitting the data in folds
+
+We use the data that is already pre-processed and we divide it in 10 random folds using `kfold_split_random`
+
+```{r prepare-folds, message=FALSE}
+# Prepare data
+roaches$fold <- kfold_split_random(K = 10, N = nrow(roaches))
+```
+
+
+## Fitting and extracting the likelihood of each fold
+
+We now loop over the 10 folds: Each time, we fit the model to all the observations except the ones belonging to the fold, but we extract the log-likelihood from the fold. The output of this loop is a list of log-likelihood matrices.
```

Comment:
"except the ones belonging to the fold, but we extract the log-likelihood from the fold. The output of this loop is a list of log-likelihood matrices."
->
"except the ones belonging to the left out fold, but we compute the log pointwise predictive densities for the left out fold. The output of this loop is a list of log predictive density matrices."

Reply:
I rephrased this a bit:

We now loop over the 10 folds. In each fold we do the following. First, we fit the model to all the observations except the ones belonging to the left-out fold. Second, we compute the log pointwise predictive densities for the left-out fold. Last, we store the predictive density for the observations of the left-out fold in a matrix. The output of this loop is a matrix of the log pointwise predictive density of all the observations.

Reply:
density -> densities

Reply:
oops